File: js/museum.js
import { createScene } from './environment.js';
import { Agent } from './agent.js';
import { Wall } from './wall.js';
import { updateAgents } from './physics.js'
import * as UTILS from './utils.js'
import * as THREE from 'three';

let agents = [];
let walls = [];
let timestep = 0.005;

const CONFIG = {
    COUNT: 150,
    RADIUS: 1,
    MAXSPEED: 10,
    MAXFORCE: 50,
    HORIZON: 1,
    K: 2,
    AVOID: 5,
    SIDESTEP: 5,    
    TRANSITIONCAP: 20,
    wAgent: 30,
    wPainting: 80,
    MINCOMFORT: 10,
    MIDCOMFORT: 20,
    MAXCOMFORT: 30,
    BLOCKED_THRESH: 1
}

let viewerCounts = {};
let paintings = [];
let viewingPoints = [];
let exitPoints = [];
let inTransition = 0;

let pickableObjects = [];
const raycaster = new THREE.Raycaster();

const agentMat = new THREE.MeshLambertMaterial({
    color: 0x00ff00
});

const { renderer, scene, camera } = createScene();
init();

// balance distance from agent and distance to painting preferences
function weightedScore(point, agent, painting) {
    return CONFIG.wAgent * point.distanceTo(agent.position) + CONFIG.wPainting * point.distanceTo(painting.position);
}

function generateViewingPosition(agent, painting) {
    return viewingPoints
        .sort((p1, p2) => weightedScore(p2, agent, painting) - weightedScore(p1, agent, painting))
        .pop();
}

function generateExitPosition(agent) {
    return exitPoints
        .sort((p1, p2) => p1.distanceTo(agent.position) - p2.distanceTo(agent.position))
        .at(0);
}

// choose least crowded painting
function choosePainting(agent) {
    let unblockedPaintings = [];

    paintings.forEach(function(painting) {
        if (painting.mesh.id != agent.getData("painting")) {
            const dir = painting.mesh.position.clone().sub(agent.position.clone()).normalize();
            raycaster.set(agent.position.clone(), dir);
            
            const intersects = raycaster.intersectObjects(pickableObjects, false);
            if (intersects.length < CONFIG.BLOCKED_THRESH) unblockedPaintings.push(painting.mesh);
        }
    });

    let minViewers = CONFIG.COUNT;
    let bestPainting = paintings[Math.floor(Math.random()*paintings.length)]; // in case all is blocked
    unblockedPaintings.forEach(function(painting) {
        if (viewerCounts[painting.id] < minViewers) {
            minViewers = viewerCounts[painting.id];
            bestPainting = painting;
        }
    });

    return bestPainting;
}

function init() {
    const wallsData = [
        [100, 25, true, new THREE.Vector3(0, 2, -50)],
        [100, 25, false, new THREE.Vector3(50, 2, 0)],
        [100, 25, false, new THREE.Vector3(-50, 2, 0)],
        [100, 25, true, new THREE.Vector3(0, 2, 50)],
    ];

    wallsData.forEach(([width, height, vertical, position]) => {
        const wall = new Wall(width, height, vertical, position);
        scene.add(wall.mesh);
        walls.push(wall);
    });

    let min = 5;
    let max = 10;
    const paintingsData = [
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), false, new THREE.Vector3(-50 + UTILS.EPSILON, 7.5, -33)],
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), false, new THREE.Vector3(-50 + UTILS.EPSILON, 7.5, -11)],
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), false, new THREE.Vector3(-50 + UTILS.EPSILON, 7.5, 11)],
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), false, new THREE.Vector3(-50 + UTILS.EPSILON, 7.5, 33)],
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), false, new THREE.Vector3(50 - UTILS.EPSILON, 7.5, -33)],
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), false, new THREE.Vector3(50 - UTILS.EPSILON, 7.5, -11)],
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), false, new THREE.Vector3(50 - UTILS.EPSILON, 7.5, 11)],
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), false, new THREE.Vector3(50 - UTILS.EPSILON, 7.5, 33)],
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), true, new THREE.Vector3(-33, 7.5, -50 + UTILS.EPSILON)],
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), true, new THREE.Vector3(-11, 7.5, -50 + UTILS.EPSILON)],
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), true, new THREE.Vector3(11, 7.5, -50 + UTILS.EPSILON)],
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), true, new THREE.Vector3(33, 7.5, -50 + UTILS.EPSILON)],
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), true, new THREE.Vector3(-33, 7.5, 50 - UTILS.EPSILON)],
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), true, new THREE.Vector3(-11, 7.5, 50 - UTILS.EPSILON)],
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), true, new THREE.Vector3(11, 7.5, 50 - UTILS.EPSILON)],
        [THREE.MathUtils.randFloat(min, max), THREE.MathUtils.randFloat(min, max), true, new THREE.Vector3(33, 7.5, 50 - UTILS.EPSILON)],
    ]

    paintingsData.forEach(([width, height, vertical, position]) => {
        const painting = new Wall(width, height, vertical, position);
        painting.mesh.material = new THREE.MeshLambertMaterial({
            color: new THREE.Color(Math.random(), Math.random(), Math.random()),
        });
        scene.add(painting.mesh);
        paintings.push(painting);
        viewerCounts[painting.mesh.id] = 0;
    });

    // precompute onlooker positions
    const p = new FastPoissonDiskSampling({
        shape: [100, 100],
        radius: 2*CONFIG.RADIUS,
        tries: 20
    }); 

    let points = p.fill();

    viewingPoints = points.map(([x, z]) => new THREE.Vector3(x - 50, 2, z - 50));
    viewingPoints = viewingPoints.filter(p => {
        if (paintings.some(painting => painting.mesh.position.distanceTo(p) < CONFIG.MINCOMFORT)) return false;
        if (!paintings.some(painting => painting.mesh.position.distanceTo(p) < CONFIG.MIDCOMFORT)) return false;
        return true;
    });

    exitPoints = points.map(([x, z]) => new THREE.Vector3(x - 50, 2, z - 50));
    exitPoints = exitPoints.filter(p => {
        if (paintings.some(painting => painting.mesh.position.distanceTo(p) < CONFIG.MIDCOMFORT)) return false;
        if (!paintings.some(painting => painting.mesh.position.distanceTo(p) > CONFIG.MAXCOMFORT)) return false;
        return true;
    });

    for (let i = 0; i < CONFIG.COUNT; i++) {
        const maxSpeed = Math.random() * (CONFIG.MAXSPEED - 5) + 5;

        agents.push(new Agent(
            i,
            0, 2, 0,
            0, 0, 0, 
            0, 0, 0,
            0, 2, 0,
            CONFIG.RADIUS, maxSpeed, CONFIG.MAXFORCE, CONFIG.HORIZON,
            CONFIG.K, CONFIG.AVOID, CONFIG.SIDESTEP                   
        ));

        const painting = paintings[Math.floor(Math.random() * paintings.length)];
        const pos = generateViewingPosition(
            agents[i],
            painting.mesh
        );
        agents[i].target = pos;
        agents[i].position = pos.clone().add(new THREE.Vector3(UTILS.EPSILON, 0, UTILS.EPSILON));
        agents[i].setData("viewingPosition", pos);
        agents[i].setData("painting", painting.mesh.id);
        agents[i].setData("state", "VIEWING");
        agents[i].setData("timer", Math.random() * 10);
        viewerCounts[painting.mesh.id]++;
    }

    agents.forEach(function(member) {
        const agentGeometry = new THREE.CylinderGeometry(member.radius, 1, 4, 16);
        const agentMaterial = new THREE.MeshLambertMaterial({
            color: 0x00ff00
        });
        const agent = new THREE.Mesh(agentGeometry, agentMaterial);
        agent.castShadow = true;
        agent.receiveShadow = true;
        scene.add(agent);
        member.setData("agent", agent);
        pickableObjects.push(agent);
    });
}

let prev = 0;
let timer = 0;

function animate(timestamp = 0) {
    requestAnimationFrame(animate);

    const delta = (timestamp - prev) / 1000;
    prev = timestamp;

    timer += delta;

    agents.forEach(function(member) {
        switch (member.getData("state")) {
            case "VIEWING":            
                // subtract time          
                member.setData("timer", member.getData("timer") - (delta * timestep / 0.005));
                
                // navigate agent to exit point if there aren't too many other exiting agents
                if (member.getData("timer") <= 0 && inTransition < CONFIG.TRANSITIONCAP) {
                    inTransition++;
                    member.target = generateExitPosition(member);
                    member.setData("state", "EXITING");
                    viewerCounts[member.getData("painting")]--;
                } else if (member.getData("timer") <= 0 && inTransition >= CONFIG.TRANSITIONCAP) {
                    member.setData("timer", Math.random() * 10);
                }
                break;
            
            case "EXITING":
                // choose crowd with small crowd for agent to navigate to after exiting crowd
                if (member.position.distanceTo(member.target) < 5) {
                    const painting = choosePainting(member);
                    viewingPoints.push(member.getData("viewingPosition"));
                    member.setData("viewingPosition", generateViewingPosition(member, painting));
                    member.target = member.getData("viewingPosition").clone();
                    member.setData("painting", painting.id);
                    member.setData("state", "WALKING");
                    viewerCounts[member.getData("painting")]++;
                }
                break;

            case "WALKING":
                // set new timer for viewer once painting is reached
                if (member.position.distanceTo(member.getData("viewingPosition")) < 5) {
                    inTransition--;
                    member.setData("timer", Math.random() * 10);
                    member.setData("state", "VIEWING");
                }
                break;
        } 
    });

    timestep = document.getElementById("timestep").value;
    document.getElementById("timestepValue").innerHTML = timestep;
    agents.forEach(function(member) {
        updateAgents(member, agents, timestep);
    });

    agents.forEach(function(agent) {
        walls.forEach(function(wall) {
            wall.collisionResolve(agent);
        });
    });

    agents.forEach(function(member) {
        member.getData("agent").position.copy(member.position);
        member.getData("agent").material = agentMat;
    });

    renderer.render(scene, camera);
};

animate()

File: py/museum.py
import pygame
from pygame.math import Vector2
from agent import Agent
from wall import Wall
from physics import update_agent
import utils
import random
from poisson_disc import Bridson_sampling
import numpy as np

CONFIG = {
    "COUNT": 150,
    "RADIUS": 3,
    "MAXSPEED": 30,
    "MAXFORCE": 150,
    "HORIZON": 3,
    "K": 3,
    "AVOID": 15,
    "SIDESTEP": 15,
    "TRANSITIONCAP": 20,
    "wAgent": 30,
    "wPainting": 80,
    "MINCOMFORT": 30,
    "MIDCOMFORT": 60,
    "MAXCOMFORT": 90,
    "BLOCKED_THRESH": 1,
}

SCREEN_WIDTH = 600
SCREEN_HEIGHT = 600
EPSILON = 0.01
TIMESTEP = 0.05

pygame.init()
screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
clock = pygame.time.Clock()

agents = []
walls = []
paintings = []
viewing_points = []
exit_points = []
viewer_counts = {}
in_transition = 0
positions = {}
frame = 0
pause = False

def weighted_score(point, agent, painting):
    return CONFIG["wAgent"] * point.distance_to(agent.position) + CONFIG["wPainting"] * point.distance_to(painting["position"])

def generate_viewing_position(agent, painting):
    if not viewing_points:
        return Vector2(300, 300)
    sorted_points = sorted(viewing_points, key=lambda p: weighted_score(p, agent, painting), reverse=True)
    best_point = sorted_points[-1]
    viewing_points.remove(best_point)
    return best_point.copy()

def generate_exit_position(agent):
    sorted_points = sorted(exit_points, key=lambda p: p.distance_to(agent.position))
    return sorted_points[0].copy() if sorted_points else Vector2(300, 300)

def choose_painting(agent):
    unblocked_paintings = []
    
    for painting in paintings:
        if painting["id"] != agent.data.get("painting"):
            unblocked_paintings.append(painting)
    
    min_viewers = CONFIG["COUNT"]
    best_painting = random.choice(paintings) if paintings else None
    
    for painting in unblocked_paintings:
        if viewer_counts[painting["id"]] < min_viewers:
            min_viewers = viewer_counts[painting["id"]]
            best_painting = painting
    
    return best_painting

walls_data = [
    (300, True, Vector2(300, 150)),
    (300, False, Vector2(450, 300)),
    (300, False, Vector2(150, 300)),
    (300, True, Vector2(300, 450)),
]

for length, vertical, pos in walls_data:
    walls.append(Wall(length, vertical, pos))

paintings_data = [
    (20, False, Vector2(((-50 + EPSILON) + 100) * 3, (-33 + 100) * 3)),
    (20, False, Vector2(((-50 + EPSILON) + 100) * 3, (-11 + 100) * 3)),
    (20, False, Vector2(((-50 + EPSILON) + 100) * 3, (11 + 100) * 3)),
    (20, False, Vector2(((-50 + EPSILON) + 100) * 3, (33 + 100) * 3)),
    (20, False, Vector2(((50 - EPSILON) + 100) * 3, (-33 + 100) * 3)),
    (20, False, Vector2(((50 - EPSILON) + 100) * 3, (-11 + 100) * 3)),
    (20, False, Vector2(((50 - EPSILON) + 100) * 3, (11 + 100) * 3)),
    (20, False, Vector2(((50 - EPSILON) + 100) * 3, (33 + 100) * 3)),
    (20, True, Vector2((-33 + 100) * 3, ((-50 + EPSILON) + 100) * 3)),
    (20, True, Vector2((-11 + 100) * 3, ((-50 + EPSILON) + 100) * 3)),
    (20, True, Vector2((11 + 100) * 3, ((-50 + EPSILON) + 100) * 3)),
    (20, True, Vector2((33 + 100) * 3, ((-50 + EPSILON) + 100) * 3)),
    (20, True, Vector2((-33 + 100) * 3, ((50 - EPSILON) + 100) * 3)),
    (20, True, Vector2((-11 + 100) * 3, ((50 - EPSILON) + 100) * 3)),
    (20, True, Vector2((11 + 100) * 3, ((50 - EPSILON) + 100) * 3)),
    (20, True, Vector2((33 + 100) * 3, ((50 - EPSILON) + 100) * 3)),
]

painting_id = 0
for size, vertical, pos in paintings_data:
    color = (random.randint(50, 255), random.randint(50, 255), random.randint(50, 255))
    painting = {
        "id": painting_id,
        "size": size,
        "vertical": vertical,
        "position": pos,
        "color": color,
    }
    paintings.append(painting)
    viewer_counts[painting_id] = 0
    painting_id += 1

points = Bridson_sampling(np.array([600, 600]), radius=2 * CONFIG["RADIUS"], k=30)
all_points = [Vector2(x, y) for x, y in points]

all_points = [p for p in all_points if 160 < p.x < 440 and 160 < p.y < 440]

for point in all_points:
    too_close = any(painting["position"].distance_to(point) < CONFIG["MINCOMFORT"] for painting in paintings)
    has_nearby = any(painting["position"].distance_to(point) < CONFIG["MIDCOMFORT"] for painting in paintings)
    
    if not too_close and has_nearby:
        viewing_points.append(point)

for point in all_points:
    too_close = any(painting["position"].distance_to(point) < CONFIG["MIDCOMFORT"] for painting in paintings)
    has_far = any(painting["position"].distance_to(point) > CONFIG["MAXCOMFORT"] for painting in paintings)
    
    if not too_close and has_far:
        exit_points.append(point)

for i in range(CONFIG["COUNT"]):
    max_speed = random.uniform(15, CONFIG["MAXSPEED"])
    painting = random.choice(paintings)
    
    start_pos = utils.get_position(175, 425, 175, 425)
    
    temp_agent = Agent(i, start_pos, CONFIG["RADIUS"], max_speed, CONFIG["MAXFORCE"],
                      CONFIG["HORIZON"], CONFIG["K"], CONFIG["AVOID"], CONFIG["SIDESTEP"])
    
    pos = generate_viewing_position(temp_agent, painting)
    
    agent = Agent(i, pos + Vector2(EPSILON, EPSILON), CONFIG["RADIUS"], max_speed, CONFIG["MAXFORCE"],
                  CONFIG["HORIZON"], CONFIG["K"], CONFIG["AVOID"], CONFIG["SIDESTEP"])
    
    agent.target = pos.copy()
    agent.data["viewing_position"] = pos
    agent.data["painting"] = painting["id"]
    agent.data["state"] = "VIEWING"
    agent.data["timer"] = random.uniform(0, 10)
    viewer_counts[painting["id"]] += 1
    
    agents.append(agent)

prev_time = pygame.time.get_ticks()

running = True
while running:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False
        elif event.type == pygame.KEYDOWN:
            if event.key == pygame.K_SPACE:
                pause = not pause

    if not pause:
        current_time = pygame.time.get_ticks()
        delta = (current_time - prev_time) / 1000.0
        prev_time = current_time

        for agent in agents:
            state = agent.data.get("state")
            
            if state == "VIEWING":
                agent.data["timer"] -= delta * TIMESTEP / 0.005
                
                if agent.data["timer"] <= 0 and in_transition < CONFIG["TRANSITIONCAP"]:
                    in_transition += 1
                    agent.target = generate_exit_position(agent)
                    agent.data["state"] = "EXITING"
                    viewer_counts[agent.data["painting"]] -= 1
                elif agent.data["timer"] <= 0:
                    agent.data["timer"] = random.uniform(0, 10)
            
            elif state == "EXITING":
                if agent.position.distance_to(agent.target) < 15:
                    painting = choose_painting(agent)
                    viewing_points.append(agent.data["viewing_position"])
                    agent.data["viewing_position"] = generate_viewing_position(agent, painting)
                    agent.target = agent.data["viewing_position"].copy()
                    agent.data["painting"] = painting["id"]
                    agent.data["state"] = "WALKING"
                    viewer_counts[agent.data["painting"]] += 1
            
            elif state == "WALKING":
                if agent.position.distance_to(agent.data["viewing_position"]) < 15:
                    in_transition -= 1
                    agent.data["timer"] = random.uniform(0, 10)
                    agent.data["state"] = "VIEWING"

        for agent in agents:
            update_agent(agent, agents, TIMESTEP)

        for agent in agents:
            for wall in walls:
                wall.collision_resolve(agent)

        frame += 1
        positions[frame] = {agent.id: (agent.position.x, agent.position.y) for agent in agents}

    screen.fill((30, 30, 30))
    
    for wall in walls:
        pygame.draw.rect(screen, (100, 100, 100), wall.rect)
    
    for painting in paintings:
        size = int(painting["size"] * 3)
        pos = (int(painting["position"].x), int(painting["position"].y))
        if painting["vertical"]:
            pygame.draw.rect(screen, painting["color"], (pos[0] - size//2, pos[1] - 2, size, 4))
        else:
            pygame.draw.rect(screen, painting["color"], (pos[0] - 2, pos[1] - size//2, 4, size))
    
    for agent in agents:
        pygame.draw.circle(screen, (0, 255, 0), (int(agent.position.x), int(agent.position.y)), int(agent.radius))

    pygame.display.flip()
    clock.tick(60)

pygame.quit()